File: algorithms/java/src/RemoveNthNodeFromEndOfList.java
/**
@author jasper
@create 2021-10-12 11:52
 */

public class RemoveNthNodeFromEndOfList {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        int len = 0;
        ListNode p = head;
        while (p!= null){
            p = p.next;
            len ++;
        }

        if (len == 1) {
            return null;
        }

        int rmNodeIndex = len - n;
        //如果删除的是头结点
        if (rmNodeIndex == 0) {
            return head.next;
        }

        //找到被删除结点的前一个结点
        p = head;
        for (int i = 0; i < rmNodeIndex - 1; i++) {
            p = p.next;
        }

        //改变指向
        p.next = p.next.next;
        return head;
    }
}

File: algorithms/python/RemoveNthNodeFromEndofList.py
# @author jasper
# @create 2021-10-12 11:52

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:
        index = []
        pos = head
        while pos is not None:
            index.append(pos)
            pos = pos.next
        ls = len(index)
        if n == ls:
            if ls > 1:
                return index[1]
            else:
                return None
        else:
            index_pos = ls - n - 1
            index[index_pos].next = index[index_pos + 1].next
            return head
